rdkPlugins/Networking/source/IPAllocator.cpp

```diff
@@ -0,0 +1,315 @@
+    if (unlink(addressFilePath.c_str()) == -1)
+    {
+        AI_LOG_WARN("failed to remove address file for container %s at %s", containerId.c_str(), addressFilePath.c_str());
+        return false;
+    }
+
+    // Remove from in-memory store
+    ContainerNetworkInfo tmp;
+    tmp.containerId = containerId;
+
+    auto itr = std::find(mAllocatedIps.begin(), mAllocatedIps.end(), tmp);
+    if (itr != mAllocatedIps.end())
+    {
+        AI_LOG_DEBUG("Deallocating IP address %s for %s", itr->ipAddress.c_str(), containerId.c_str());
+
+        // Remove allocation and add back to the unallocated queue for re-use
+        mAllocatedIps.erase(itr);
+        mUnallocatedIps.push(itr->ipAddressRaw);
+    }
+
+    AI_LOG_FN_EXIT();
+    return true;
+}
+
+/**
+ * @brief Retrieves the networking information (veth, ip) for a given container
+ *
+ * @param[in]   containerId     Name of the container to retrieve the information about
+ * @param[out]  networkInfo     Struct to store the container network info in
+ *
+ * @return True on success, false on failure
+ */
+bool IPAllocator::getContainerNetworkInfo(const std::string &containerId, ContainerNetworkInfo &networkInfo) const
+{
+    const std::string filePath = ADDRESS_FILE_DIR + mUtils->getContainerId();
+    return getNetworkInfo(filePath, networkInfo);
+}
+
+/**
+ * @brief Retrieves the networking information (veth, ip) from a file from the store
+ *
+ * @param[in]   filePath        Path to the file to parse for information about the container network
+ * @param[out]  networkInfo     Struct to store the container network info in
+ *
+ * @return True on success, false on failure
+ */
+bool IPAllocator::getNetworkInfo(const std::string &filePath, ContainerNetworkInfo &networkInfo) const
+{
+    AI_LOG_FN_ENTRY();
+
+    // Parse the file
+    const std::string addressFileStr = mUtils->readTextFile(filePath);
+    if (addressFileStr.empty())
+    {
+        AI_LOG_ERROR_EXIT("failed to get IP address and veth name assigned to container from %s",
+                          filePath.c_str());
+        return false;
+    }
+
+    const std::string ipStr = addressFileStr.substr(0, addressFileStr.find("/"));
+    const in_addr_t ip = std::stoi(ipStr);
+
+    // check if string contains a veth name after the ip address
+    if (addressFileStr.length() <= ipStr.length() + 1)
+    {
+        AI_LOG_ERROR_EXIT("failed to get veth name from %s", filePath.c_str());
+        return false;
+    }
+
+    networkInfo.containerId = basename(filePath.c_str());
+    networkInfo.ipAddressRaw = ip;
+    networkInfo.ipAddress = ipAddressToString(htonl(ip));
+    networkInfo.vethName = addressFileStr.substr(ipStr.length() + 1, addressFileStr.length());
+
+    AI_LOG_FN_EXIT();
+    return true;
+}
+
+/**
+ * @brief Synchronise the in-memory pool of allocated IPs with the disk store
+ *
+ * @return True on success
+ */
+bool IPAllocator::updateFromStore()
```

Comment:
I don't like how this method looks like. You are not clearing mAllocatedIps on beginning and using emplace_back which suggest that you expect that someone can run "updateFromStrore" while Allocator is already running, but at the same time you are not updating mUnallocatedIps. So I am not sure what is the desired efect here. If you would like to have updates while running please update also the mUnallocatedIps (but this would make this so much like the constructor itself). If this is helper function for the constructor and shouldn't be running after Allocator was constructed then clean mAllocatedIps at the beggining (in case someone us it like that) and maybe change the name of this method.